libs/src/api/user.js

```diff
@@ -683,6 +684,58 @@ class Users extends Base {
   _cleanUserMetadata (metadata) {
     return pick(metadata, USER_PROPS.concat('user_id'))
   }
+
+  // Perform replica set update
+  // Conditionally write to UserFactory contract, else write to UserReplicaSetManager
+  // This behavior is to ensure backwards compatibility prior to contract deploy
+  async _updateReplicaSetOnChain (userId, creatorNodeEndpoint) {
+    // Attempt to update through UserReplicaSetManagerClient if present
+    if (!this.contracts.UserReplicaSetManagerClient) {
+      await this.contracts.initUserReplicaSetManagerClient()
+    }
+    // If still uninitialized, proceed with legacy update - else move forward with new contract update
+    if (!this.contracts.UserReplicaSetManagerClient) {
+      const { txReceipt: updateEndpointTxReceipt } = await this.contracts.UserFactoryClient.updateCreatorNodeEndpoint(
+        userId,
+        creatorNodeEndpoint
+      )
+      return updateEndpointTxReceipt
+    }
+    let primaryEndpoint = CreatorNode.getPrimary(creatorNodeEndpoint)
+    let secondaries = CreatorNode.getSecondaries(creatorNodeEndpoint)
+    let [primarySpID, secondary1SpID, secondary2SpID] = await Promise.all([
+      this._retrieveSpIDFromEndpoint(primaryEndpoint),
+      this._retrieveSpIDFromEndpoint(secondaries[0]),
+      this._retrieveSpIDFromEndpoint(secondaries[1])
+    ])
+    // Update in new contract
+    let tx = await this.contracts.UserReplicaSetManagerClient.updateReplicaSet(
+      userId,
+      primarySpID,
+      [secondary1SpID, secondary2SpID]
+    )
+    return tx
+  }
+
+  // Retrieve cached value for spID from endpoint if present, otherwise fetch from eth web3
+  // Any error in the web3 fetch will short circuit the entire operation as expected
+  async _retrieveSpIDFromEndpoint (endpoint) {
+    let cachedSpID = getSpIDForEndpoint(endpoint)
+    let spID = cachedSpID
+    if (!spID) {
+      let spEndpointInfo = await this.ethContracts.ServiceProviderFactoryClient.getServiceProviderInfoFromEndpoint(
+        endpoint
+      )
+      // Throw if this spID is 0, indicating invalid
+      spID = spEndpointInfo.spID
+      if (spID === 0) {
```

Comment:
@SidSethi tested with an invalid endpoint and the value is zero so this should be covered

Reply:
perfect